Some fixes that help with useability.

TheGreatCodeholio opened the pull request:
File formatting, log formatting fixed in some files. Fixed typo requets to request

Fixed Sending Images included option allow_forward, allow_save

Added Random GIF option picks a random GIF from search.

Fixed parsing different types of group/chat messages. So you can tell where a image/sticker/gif came from.

Modified echo bot to include on temp ban received and on pong received as examples.

Added getting gateway node based on kik APK version. Instead of using an old gateway.

Added method to catch temp bans and on pong to echo bot example

TheGreatCodeholio committed:
File formatting, log formatting fixed in some files.

Fixed Sending Images included option allow_forward, allow_save

Added Random GIF option picks a random GIF from search.

Fixed parsing different types of group/chat messages. So you can tell where a image/sticker/gif came from.

Modified echo bot to include on temp ban received and on pong recieved as examples.

Added getting gateway node based on kik APK version. Instead of using an old gateway.

Added method to catch temp bans

TheGreatCodeholio committed:
Added example bot as a more robust and advanced example or echo bot.

Fixed a bug in image uploading from last commit. Added handling retries for image uploading if there is an error. Will retry 5 times max at 2 second intervals.

Changed generating credentials from yaml to json. Seems to make more sense to be in json/dict instead of as yaml.

Added a ServiceRequestError type for handling returned errors when requesting something from kik servers. Peers or Roster requests can sometimes give a 503 unavailable error. With this on_service_error you should be able to match message_id with requests made and remake requests.

Lastly typo fixes

tomer8007 commented on kik_unofficial/datatypes/xmpp/chatting.py:
```diff
@@ -60,43 +60,56 @@ class OutgoingChatImage(XMPPElement):
+            f'</hashes>'
+            f'<images>'
+            f'<preview>{self.parsed_image["base64"]}</preview>'
+            f'<icon>iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAAAAXNSR0IArs4c6QAAAARzQklUCAgICHwIZIgAAAMpSURBVGiB7ZdNSBRhGMd/M1NJhZaGlRHmIRULsmN0sGsXqUOnoEMQhF07WRB0SPuAkjBCSTzUIfDS0sEgsCw6lVQmmbrq+oHa5k7q6urqzkyHSXTdHZuvnF3Z32n2fR/m/f/3fd73eUZA04SydxxWFao1OAPkkdrIAvhEidruCvxCWbtWrCi8BEq9VmaRHkmiUlQVqkk/8QClqkK1UNKmhUj9tDFCFklf8QB5otcKnJIx4DUZA16zxWjidD5cL4a9WRspJ5FgFGr6oPVX8nnDHUgF8aBruFZsPG9 ...
```
What's this hardcoded `icon`?

TheGreatCodeholio replied:
I don't remember why that is there exactly. I do know if it is not there the image does not display properly. You can not click on the image to zoom in.

tomer8007 replied:
What image does this icon decode to?

TheGreatCodeholio replied:
Here is what it decodes to. 
![image](https://github.com/tomer8007/kik-bot-api-unofficial/assets/27271698/8e356b87-55c2-4f95-bae7-1f3f7e7426f7)

tomer8007 commented on kik_unofficial/datatypes/xmpp/login.py:
```diff
@@ -49,19 +49,19 @@ def serialize(self) -> bytes:
                 f'<device-id>{self.device_id}</device-id>'
                 '<install-referrer>utm_source=google-play&amp;utm_medium=organic</install-referrer>'
-                '<operator>unknown</operator>'
+                '<operator>310260</operator>'
```
Why did you change it to this operator?

TheGreatCodeholio replied:
I changed the operator, make and model to be more consistent with an actual phone. I don't know if this matters, but my thought process is if you are going to emulate a phone. Adding correct information will help the bot/api not stand out amongst other unofficial clients. Do less likely to be targeted that way for a ban wave.

tomer8007 replied:
OK

tomer8007 commented:
Thank you, this looks good

TheGreatCodeholio committed:
Added Video Messages
You can now send a video message. example self.client.send_chat_video(chat_message.from_jid, "var/test_video.mp4", auto_play=True, looping=True)

Video must be in mp4 format and less than 15MB in size. Otherwise KIK servers will reject it.

Updated requirements.txt to include new requirements for processing a video file.

Reworked the Random GIF code to make it more robust.

Added method to send a saved GIF. Useful if you want your bot to save a GIF to send later. Such as in a substitute or trigger. You woudl save the GIF to a json file, and then when trigger/substitution is called send the saved_gif_image. Example self.client.send_saved_gif_image(chat_message.from_jid, "var/test_gif.json")

Example bot added more command examples in PM now there is image, video, GIF, Random GIF (gifr), Saved GIF (gifs), ping

TheGreatCodeholio commented:
Added another commit ^ just so you are aware :)

TheGreatCodeholio committed:
fixed setup.py so it includes new requirements

TheGreatCodeholio committed:
fixed setup.py so it includes new requirements

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -204,6 +204,34 @@ def send_chat_image(self, peer_jid: str, image_file, allow_forward=False, allow_
+    def send_chat_video(self, peer_jid: str, video_file, forward=True, save=True,
+                        auto_play=False, muted=False, looping=False):
+        """
+        Sends an image chat message to another person or a group with the given JID/username.
```
You probably mean "Sends a video message"

tomer8007 commented:
Nice! :-) Video messages are a highly requested feature.
How about calling it `VideoMessage`/`video_message` instead of `ChatVideo`\ `chat_video`?
Also let me know when you're done with the commits here so I can merge.

TheGreatCodeholio committed:
ChatVideo\ chat_video to VideoMessage/video_message

TheGreatCodeholio commented:
Okay made that change and I am done with commits for now.

tomer8007 commented:
It still says "Sends an image chat message" inside `send_video_messaage`

TheGreatCodeholio committed:
added a globals.py file to hold global variables.

Made it so you can utilize the API logger in your bot script. example in example_bot.py This way there is no need two logger instances.

Made on_connect generate a new device id only if one is not supplied. This is helpful for inital login with password even if you know your kik node name.

TheGreatCodeholio committed:
Fixed send_video_message docstring

TheGreatCodeholio commented:
Fixed both the docstring for video message and added a new file globals to store global variables in. Currently utilized for the logger. So user can user the logger instance in the API for their bot script to unify them. Example in example_bot. Think two other small changes as well.

tomer8007 commented:
Why do we need the ability to set `logger_name` in the constructor?

TheGreatCodeholio commented:
> Why do we need the ability to set `logger_name` in the constructor?

My thought process was a long of lines of my own bot I have created. I have multiple files and across those file I get the with `getLogger(log_name)`

Adding the ability to set the log name in the constructor allows the user to essentially piggy back their bot script logs on top of the logging classes I added to the API. Being able to specify the name of the logger, is somewhat essential in that process.

tomer8007 commented:
Let's just move the `log_level` argument to be before the `logger_name`, because it's more likely to be changed.

TheGreatCodeholio committed:
changed the formatting for logs slightly. - for error and # for critical to match closer to original format.

moved variables for constructor log_level is not before logger_name

generating device_id now happens in init if device_id is None.

If authentication with kik_node fails it now sets kik_node to None and retries to connect with username/password

wait for messages now only prints one time on first iteration instead of spamming logs every second.

TheGreatCodeholio commented:
I fixed that and I made a couple small changes to improve flow. 

If trying to authenticate and it fails. It now sets kik_node to None and retries connection with user/password

Device_id generation moved to __init__ only generates if device_id is None. This save some hassle, from trying to log in without node to get an authenticated session. Prior device_id was changing every time even if it was given.

tomer8007 commented:
> If trying to authenticate and it fails. It now sets kik_node to None and retries connection with user/password

I hope you don't mean that it fails now, only that you added failure handling code

TheGreatCodeholio commented:
> > If trying to authenticate and it fails. It now sets kik_node to None and retries connection with user/password
> 
> I hope you don't mean that it fails now, only that you added failure handling code

Correct, I have just added error handling. 

If self.kik_node is provided, it tries to authenticated using existing session. However.... if that fails the previous behavior was to exit. It now forces a login using username/password

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -44,20 +47,23 @@ def __init__(self, callback: callbacks.KikClientCallback, kik_username: str, kik
-            self.log = set_up_basic_logging(log_level=log_level, logger_name="kik_unofficial", log_file_path=log_file_path)
+
+        globals.LOGGER_NAME = logger_name
+
```
I think it's a little weird that you set a global that appears to be constant here in the constructor. You can probably avoid it

TheGreatCodeholio replied:
Its a variable from the globals.py file. Doing it this way allows other files in the project to use whatever variables in it as a global variable. 

I may be wrong, and it might be able to be avoided, but logging in other files has to be able to get the correct logger by name when set. 

`log = logging.getLogger(globals.LOGGER_NAME)`

tomer8007 replied:
That's okay, but having multiple `KikClient`s will set `globals.LOGGER_NAME` multiple times which might be confusing.

tomer8007 replied:
It's either global or per `KikClient`

TheGreatCodeholio replied:
I see what you mean. I did not account for that. Is there another solution that might work?

tomer8007 replied:
I don't think there is a need to unify the API logs and the bot logs. If you think it's needed, you can have `logger_name` as a member of `KikClient` and have a function to export that for whoever.

TheGreatCodeholio replied:
I have removed the globals and setting the logger_name.

TheGreatCodeholio committed:
removed globals.py removed setting up logger name in constructor

tomer8007 commented on kik_unofficial/datatypes/xmpp/auth_stanza.py:
```diff
@@ -8,9 +8,8 @@
-from kik_unofficial.utilities import globals
 
-log = logging.getLogger(globals.LOGGER_NAME)
+log = logging.getLogger("kik_unofficial")
```
I think a better name for the variable `log` would be `logger`

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -55,9 +53,7 @@ def __init__(self, callback: callbacks.KikClientCallback, kik_username: str, kik
-        globals.LOGGER_NAME = logger_name
-
-        self.log = set_up_basic_logging(log_level=log_level, logger_name=logger_name, log_file_path=log_file_path)
+        self.log = set_up_basic_logging(log_level=log_level, logger_name="kik_unofficial", log_file_path=log_file_path)
```
Also here how about `self.logger`?

TheGreatCodeholio committed:
changed logger name from log to logger

added NoneType check for uris in incoming GIF messages. Seems some GIFs are sent with URI in a weird place. Assuming custom clients do that.

TheGreatCodeholio commented:
I changed log to logger for both instances. self.log and log = so now they are self.logger and logger respectively.

Added URI check for NoneType for incoming GIF messages. Odd issues with receiving some GIF messages. Not sure if its a custom client thing or a parsing issue with BeautifulSoup

EDIT:
It is a parsing issue. It happens when a GIF or Video is sent in two parts. Here is a debug log sample. Notice the second portion is not considered an addition to the first portion is is just considered a "document" by the api because there is no type data. 

```
2023-10-18 17:24:55 DEBUG: [^] [Kik Connection] Received raw data: b'<message to="lucky_bot_69_yx3@talk.kik.com" type="groupchat" xmlns="jabber:client" from="be76ygcvjr2jiydljwvsoqnmmpzi37b4et25hfi5pv6soveqvfpq_a@talk.kik.com" id="caf8030a-1776-47e7-bbf6-a0ff7f43d5ab" cts="1697664294288"><kik qos="true" timestamp="1697664294276" hop="true" push="true" app="chat"/><pb/><content v="2" app-id="com.kik.ext.gif" id="9201d717-b8e0-4d5d-9492-dc826acc85a2"><strings><app-name>GIF</app-name><layout>video</layout><allow-forward>true</allow-forward><disallow-save>true</disallow-save><video-should-autoplay>true</video-should-autoplay><video-should-loop>true</video-should-loop><video-should-be-muted>true</video-should-be-muted></strings><images><icon>iVBOR(...)==</icon><preview>/9j/4A(...)fs5X3r'
2023-10-18 17:24:55 DEBUG: [^] [Kik Connection] Received raw data: b'mpCZJAP(...)/2Q==</preview></images><uris><uri priority="0" file-content-type="video/mp4" type="video">https://media.tenor.com/wH9jFSjtfl0AAAPo/sharon-den-adel-whistling.mp4</uri><uri priority="1" file-content-type="video/webm" type="video">https://media.tenor.com/wH9jFSjtfl0AAAPs/sharon-den-adel-whistling.webm</uri><uri priority="0" file-content-type="video/tinymp4" type="video">https://media.tenor.com/wH9jFSjtfl0AAAP1/sharon-den-adel-whistling.mp4</uri><uri priority="1" file-content-type="video/tinywebm" type="video">https://media.tenor.com/wH9jFSjtfl0AAAP3/sharon-den-adel-whistling.webm</uri><uri priority="0" file-content-type="video/nanomp4" type="video">https://media.tenor.com/wH9jFSjtfl0AAAP2/sharon-den-adel-whistling.mp4</uri><uri priority="1" file-content-type="video/nanowebm" type="video">https://media.tenor.com/wH9jFSjtfl0AAAP4/sharon-den-adel-whistling.webm</uri></uris></content><request xmlns="kik:message:receipt" d="true" r="true"/><g jid="1100251194664_g@groups.kik.com"/><alias-sender>be76ygcvjr2jiydljwvsoqnmmpzi37b4et25hfi5pv6soveqvfpq_a@talk.kik.com</alias-sender></message>'
Exception in thread Thread-661 (_on_new_data_received):
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ian/PycharmProjects/lucky_bot/venv/lib/python3.11/site-packages/kik_unofficial/client.py", line 621, in _on_new_data_received
2023-10-18 17:24:55 WARNING: [!] [Thread-662 (_on_new_data_received)] Unknown element type: [document]
    self._handle_xmpp_message(xml_element)
  File "/home/ian/PycharmProjects/lucky_bot/venv/lib/python3.11/site-packages/kik_unofficial/client.py", line 719, in _handle_xmpp_message
    self._handle_kik_event(xmpp_message)
  File "/home/ian/PycharmProjects/lucky_bot/venv/lib/python3.11/site-packages/kik_unofficial/client.py", line 732, in _handle_kik_event
    xmlns_handlers.GroupXMPPMessageHandler(self.callback, self).handle(xmpp_element)
  File "/home/ian/PycharmProjects/lucky_bot/venv/lib/python3.11/site-packages/kik_unofficial/xmlns_handlers.py", line 155, in handle
    self.callback.on_gif_received(IncomingGifMessage(data))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ian/PycharmProjects/lucky_bot/venv/lib/python3.11/site-packages/kik_unofficial/datatypes/xmpp/chatting.py", line 502, in __init__
    self.uris = [self.Uri(uri) for uri in data.content.uris]
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not iterable
```

tomer8007 commented:
Any more changes you want to commit?

TheGreatCodeholio commented:
Nope no more changes for now.

TheGreatCodeholio committed:
fixed KikConnection Class so that partial data is parsed correctly. such as larger gifs and videos that are sent in multiple packets.

TheGreatCodeholio commented:
I lied, I had one more to fix the above multi packet issue.

TheGreatCodeholio committed:
fixed parsing uri now that we have multi data packet GIFS

tomer8007 commented:
Your latest changes with the parsing issue look good, I'm waiting for your final commit.

TheGreatCodeholio commented:
No more commits for this one. :)

TheGreatCodeholio commented:
I am making an experimental commit. That I may revert if it doesn't work as I want it to. I did some more tinkering with received data.

TheGreatCodeholio committed:
modified parsing new data packets to see if we can handle them a bit better. Hopefully able to handle out of non sequential and out of order data packets.

TheGreatCodeholio commented:
Changes worked like a charm for data packets received. It was having issues due to data collecting in partial data. With this new way if something doesn't match it sits in buffer. Looking in to ways to remove items from buffer after a timeout so we don't have memory leaks.

TheGreatCodeholio committed:
modified data_received further to allow for a timed buffer so we can empty any data not matched that is older than 1 minute.

Packet Data now is a list of tuples with timestamp and the data.

TheGreatCodeholio committed:
handle pong differently instead of an open re search make it match the ack message

TheGreatCodeholio committed:
added debug information to cleanup method

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -865,38 +868,38 @@ def data_received(self, data: bytes):
+        elif re.search(b'<pong/>', data):
+            self.loop.call_soon_threadsafe(self.api._on_new_data_received, data)
+        elif not self.is_multi_packet(data):
+            self.loop.call_soon_threadsafe(self.api._on_new_data_received, data)
```
Let's add a comment that says that this is the usual flow (no multi packet, no pong, not empty)

TheGreatCodeholio replied:
Added

tomer8007 replied:
Let's just put that after the `elif` line...
EDIT: Nevermind, I like your new comments

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -868,7 +867,8 @@ def data_received(self, data: bytes):
         if data == b" ":
             self.loop.call_soon_threadsafe(self.api._on_new_data_received, data)
-        elif re.search(b'<pong/>', data):
+        # Sometimes Pong comes back in an ack message, handle those.
```
Are there cases where `pong` comes back without an ack message?

TheGreatCodeholio replied:
Yes it can come both ways. I added better handling to this to search specifically for just ack packets.

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -913,10 +913,18 @@ def is_multi_packet(self, data: bytes):
     def cleanup_buffer(self):
         # This method is called after the cleanup_interval to remove outdated data from the buffer
         current_time = time.time()
+        cleaned_count = 0  # Initialize a counter for cleaned messages
```
You initiate `cleaned_count` here to zero, but a few lines later you initiate it to something else?

TheGreatCodeholio replied:
I set it to 0 at the beginning to clear the last value it had.

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -913,10 +913,18 @@ def is_multi_packet(self, data: bytes):
+
         self.logger.debug("Cleaning up partial data buffer")
         self.data_buffer = [(ts, data) for (ts, data) in self.data_buffer if current_time - ts <= self.cleanup_interval]
+
```
It sounds like `data_buffer` is a buffer, but in fact it's an array of timestamps and buffers. Let's make this clear

TheGreatCodeholio replied:
Changed to data_array

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -913,10 +913,18 @@ def is_multi_packet(self, data: bytes):
         self.data_buffer = [(ts, data) for (ts, data) in self.data_buffer if current_time - ts <= self.cleanup_interval]
+
+        # Calculate the number of messages cleaned
+        cleaned_count = len(self.data_buffer) - cleaned_count
```
Do I miss anything or that `cleaned_count` will always be `len(self.data_buffer)`?

TheGreatCodeholio replied:
You are correct, my logic was wrong. I have fixed it so it now sets cleaned_count to len of the array then cleans it. Then subtracts cleaned_count from the length after its been cleaned.

TheGreatCodeholio committed:
modifications for KikConnection Class. Changed name from data_buffer to data_array.

added comment for packet logic.

changed the way we parse for pong with an ack.

TheGreatCodeholio committed:
Modified the logic for incoming packet cleanup. Clean count is now correct.

TheGreatCodeholio committed:
fixed video resolution by getting a thumbnail the same size as the original video.

TheGreatCodeholio committed:
reworked the multi-packet proccessing further to be more robust so future edge cases can be handled if need be.

TheGreatCodeholio committed:
Outgoing GIF was faaling to parse some GIFs due to the non ascii formatting. Fixed by changing to utf-8.

Added exception for unicodedecode error

TheGreatCodeholio committed:
Added removing ack messages since we don't do anything with them.

Some messages come with ack messages before or after as one packet. That is an issue when parsing. So we now remove ack messages if they exist.

TheGreatCodeholio committed:
after removing ack we still have an empty bytes string. sending that thgouh on_new_data causes unknown element because its empty.

Fixed by not sending it through.

TheGreatCodeholio committed:
split messages incase there is more than one data message inside a packet.

TheGreatCodeholio committed:
fixed the regex to parse elements from a group of data packets

TheGreatCodeholio committed:
Added a way to exit from client.

added close to disconnect. If set to True will stop the main loop and wait_for_messages with out a delay.

TheGreatCodeholio committed:
changed the variable for disconnect to exiting. This method is called by other parts of the connection to disconnect and reconnect, but if exiting is True it should close all connections and stop the loop.

TheGreatCodeholio committed:
Changed cleanup message to from warning to debug.

TheGreatCodeholio committed:
Added requested enhancement, returning alias jid with add friend request.

references -  Add friend should return the public id it was called on in its callback responce [enhancement] #182

TheGreatCodeholio committed:
set alias_jid on peer info response

TheGreatCodeholio committed:
allow all friend requests to go through instead of ignoring non alias_jid

TheGreatCodeholio committed:
added a dm_privacy value to GroupMember object to see if the user has their direct messages turned off.

added alias_jid to User/Peer So when we get friend request data back we can match users better.

Fixed parsing xmpp elements. Now all elements are making it through undisturbed.

Removed checking for ack messages and removing infavor of just splitting all xmpp messages and letting parser handle ack.

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -856,9 +856,21 @@ def data_received(self, data: bytes):
-        # Handle 'ack' packets with 'pong'
-        if re.search(b'<ack id="[^"]*"/><pong/>', data):
-            self.loop.call_soon_threadsafe(self.api._on_new_data_received, data)
+        # remove ack packets from data Sometimes they can surround other data and make it useless.
```
A better solution to this problem would be to change `on_new_data_received` code to iterate the received XML elements and ignore the `ack`s. If you choose to use the regex aproach, I think we can make a shorter comment than this one to show the problem. Something like `<ack id=.... /><iq id=...></iq>`

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -890,6 +894,26 @@ def data_received(self, data: bytes):
             self.cleanup_task = self.loop.call_later(self.cleanup_interval, self.cleanup_buffer)
 
+    @staticmethod
+    def split_elements(data_bytes):
```
Again, this is something that `BeautifulSoup` already does, so if you want to be more elegant we can handle multiple XML elements inside `on_new_data_received` or in a different function

tomer8007 replied:
Also, what about future elements that are not `k`, `pong`, `message`, or `stc`?

TheGreatCodeholio replied:
I have gone over this again after using it for a few weeks on my bot. There were some errors, but only because it was not catching a second <message> element attached to the second part of multi-packet data. 

I made some changes so that the split happens after we match messages for multi-packet messages. If they are not multi packet we then split up in to multiple message elements and send to the loop.

BeautifulSoup can do something similar yes. I am unsure how to implement it to split the xml elements in to the correct groups so i am using regex. Currently the elements it is looking for are the only elements that get sent from the KIK servers. If there are more in the future we can add different regex.

tomer8007 replied:
Yes this is better

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -569,15 +573,17 @@ def change_email(self, old_email, new_email):
         return self._send_xmpp_element(account.ChangeEmailRequest(self.password, old_email, new_email))
 
-    def disconnect(self):
+    def disconnect(self, close=False):
```
Since the purpose of this function is to close the connection the meaning of the argument `close` is not really clear

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -573,17 +573,17 @@ def change_email(self, old_email, new_email):
         return self._send_xmpp_element(account.ChangeEmailRequest(self.password, old_email, new_email))
 
-    def disconnect(self, close=False):
+    def disconnect(self, exiting=False):
```
OK, let's call this argument `is_exiting`

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -573,17 +573,17 @@ def change_email(self, old_email, new_email):
         self.is_expecting_connection_reset = True
 
-        if close:
+        if exiting:
```
And let's add a comment that explains what happens in this case

TheGreatCodeholio replied:
I have changed this to is_exiting and added a comment and doc string

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -320,10 +321,17 @@ def request_info_of_users(self, peer_jids: Union[str, List[str]]):
-        return self._send_xmpp_element(roster.AddFriendRequest(peer_jid))
+        message = roster.AddFriendRequest(peer_jid)
+        if "_a@talk" in peer_jid:
+            self.alias_jid_cache[message.message_id] = peer_jid
```
So `alias_jid_cache` is more of a mapping between sent friend requests and their associated alias JIDs. Maybe we can save the whole `AddFriendRequest`s as an array or make a better name for this variable?

TheGreatCodeholio replied:
I changed the name of this mapping to friend_request_mapping. 

Its only purpose is to hold the message_id of a friend request. So that when the friend request is returned we can add the alias_jid it was requested on to the Peer Info Response message. It is deleted once it is matched to a response.

tomer8007 commented on kik_unofficial/client.py:
```diff
@@ -914,15 +896,31 @@ def split_elements(data_bytes):
+            for match in matches:
+                data_bytes = data_bytes.replace(match, b'', 1)
+
+        complete_message_pattern = b'<message .*?>.*?</message>'
```
If you split the elements and then try to identify multi-packets, I think it is going to be error prone. It will be better to first make sure we have the complete XML, and then parse it, like you did previously.

TheGreatCodeholio committed:
none type check for iq element attributes

tomer8007 commented:
I left some comments on your previous commits

TheGreatCodeholio committed:
added a fix for video thumbnails.

TheGreatCodeholio committed:
Automatic Video Thumbnails for video messages. using imageio_ffmpeg (requires ffmpeg)

Changed some variable names such as is_exiting for disconnect method.

Modified incoming message parsing to better parse xml messages.
We now combine multi-packet data before trying to split into multiple messages.This will help grab elements that come at the end of multi-packet messages.

Example:
b'<message from="oijfrm26pszo3kfnxyn4ikoxmburvrdhuk4tv3je33u4yvy45atq_a@talk.kik.com" to="***@talk.kik.com" xmlns="kik:groups" cts="1699845983321" type="groupchat" id="754efabc-49b4-4986-b52f-86950f8e20a8"><pb/><kik push="true" qos="true" app="chat" timestamp="1699845983321" hop="true"/><request xmlns="kik:message:receipt" d="true" r="true"/><content app-id="com.kik.ext.gallery" v="2" id="76a9f5f8-cfeb-4a34-8feb-a20ff34a40d7"><strings><app-name>Gallery</app-name><file-size>234693</file-size><allow-forward>true</allow-forward><file-name>***</file-name><file-url>***</file-url></strings><extras/><hashes><sha1-scaled>*</sha1-scaled><blockhash-scaled>*</blockhash-scaled><sha1-original>*</sha1-original></hashes><images><preview>***</preview><icon>***</icon></images><uris/></content><g jid="************_g@groups.kik.com"/><alias-sender>oijfrm26pszo3kfnxyn4ikoxmburvrdhuk4tv3je33u4yvy45atq_a@talk.kik.com</alias-sender></message><message from="oijfrm26pszo3kfnxyn4ikoxmburvrdhuk4tv3je33u4yvy45atq_a@talk.kik.com" to="******@talk.kik.com" xmlns="kik:groups" cts="1699845983333" type="groupchat" id="ac339c84-4cf3-4ca3-8fba-ab2df193c9e9"><body>I\'m back! I see we are having meme time hehehe</body><pb/><preview>I\'m back!...</preview><kik push="true" qos="true" app="chat" timestamp="1699845983333" hop="true"/><request xmlns="kik:message:receipt" d="true" r="true"/><ri/><g jid="***@groups.kik.com"/><alias-sender>oijfrm26pszo3kfnxyn4ikoxmburvrdhuk4tv3je33u4yvy45atq_a@talk.kik.com</alias-sender></message>'

tomer8007 commented:
Thanks for the fixes and the enhancements, this is better, but I really think that parsing XML with regex is a [bad](https://stackoverflow.com/questions/8577060/why-is-it-such-a-bad-idea-to-parse-xml-with-regex) [idea](https://flapenguin.me/xml-regex) for the long term. It's probably worth figuring out how to do this in `BeautifulSoup`, it should be easy. Don't you think?

bluemods commented:
> Thanks for the fixes and the enhancements, this is better, but I really think that parsing XML with regex is a [bad](https://stackoverflow.com/questions/8577060/why-is-it-such-a-bad-idea-to-parse-xml-with-regex) [idea](https://flapenguin.me/xml-regex) for the long term. It's probably worth figuring out how to do this in `BeautifulSoup`, it should be easy. Don't you think?

The only reliable way to parse an XML stream (XMPP) that I found is pull parsing, where the input stream is fed directly into the parser. Regex / manual checks for where a tag might be closed is error prone, inefficient, and will eventually break.
You can see my implementation in Go here: https://github.com/bluemods/kik-go-proxy/blob/8c434d8d8a27e7d0bef838ba5597e07635f5187a/node.go#L77C12-L77C12

The problem with integrating it into this repo could mean massive refactoring, but it's probably worth it to do.

tomer8007 commented:
@TheGreatCodeholio You now have to work with @bluemods new XML parser in #241. Looking forward for the video sending support.

TheGreatCodeholio committed:
Update device_configuration.py
